File: backend/src/kuuna_backend/jobs/todo_export.py
from __future__ import annotations

import logging
from datetime import datetime, timezone

import httpx
from sqlalchemy import select

from kuuna_backend.config.settings import get_settings
from kuuna_backend.db.models import Todo, TodoStatus
from kuuna_backend.integrations.postgres import get_db_session

logger = logging.getLogger(__name__)


def export_pending_todos_job(limit: int = 100, provider_group_id: str | None = None) -> int:
    settings = get_settings()
    if not settings.todo_export_enabled:
        logger.info("todo_export_disabled")
        return 0

    if not settings.todo_export_webhook_url:
        logger.warning("todo_export_webhook_missing")
        return 0

    db = get_db_session()
    exported_count = 0
    try:
        stmt = (
            select(Todo)
            .where(
                Todo.exported_at.is_(None),
                Todo.status.in_([TodoStatus.OPEN, TodoStatus.IN_PROGRESS]),
            )
            .order_by(Todo.created_at.asc())
            .limit(limit)
        )
        if provider_group_id:
            stmt = stmt.where(Todo.provider_group_id == provider_group_id)

        todos = db.scalars(stmt).all()
        for todo in todos:
            todo.export_attempt_count += 1
            try:
                external_ref = _post_todo(settings.todo_export_webhook_url, todo)
            except Exception as exc:
                todo.last_export_error = str(exc)[:1000]
                logger.warning(
                    "todo_export_failed",
                    extra={
                        "todo_id": str(todo.id),
                        "provider_group_id": todo.provider_group_id,
                        "error": str(exc),
                    },
                )
                continue

            todo.exported_at = datetime.now(timezone.utc)
            todo.external_ref = external_ref
            todo.last_export_error = None
            exported_count += 1

        db.commit()
        return exported_count
    except Exception:
        db.rollback()
        logger.exception("todo_export_job_failed")
        raise
    finally:
        db.close()


def _post_todo(webhook_url: str, todo: Todo) -> str | None:
    settings = get_settings()
    response = httpx.post(
        webhook_url,
        json={
            "id": str(todo.id),
            "provider_group_id": todo.provider_group_id,
            "message_id": str(todo.message_id) if todo.message_id else None,
            "agent_run_id": str(todo.agent_run_id) if todo.agent_run_id else None,
            "title": todo.title,
            "description": todo.description,
            "status": todo.status.value,
            "priority": todo.priority.value,
            "due_at": todo.due_at.isoformat() if todo.due_at else None,
            "created_at": todo.created_at.isoformat(),
            "updated_at": todo.updated_at.isoformat(),
        },
        timeout=settings.todo_export_timeout_seconds,
    )
    response.raise_for_status()

    payload = response.json() if response.content else {}
    if isinstance(payload, dict):
        external_ref = payload.get("external_ref") or payload.get("id")
        if isinstance(external_ref, str) and external_ref:
            return external_ref
    return None


def main() -> None:
    export_pending_todos_job()


if __name__ == "__main__":
    main()

File: services/backend-ts/src/jobs/todo-export.ts
import { and, asc, eq, inArray, isNull } from "drizzle-orm";

import { getSettings } from "../config.js";
import type { DbLike } from "../db/client.js";
import { todos } from "../db/schema.js";
import { logger } from "../logging.js";

type HttpClient = (url: string, init: RequestInit) => Promise<Response>;

type TodoRow = typeof todos.$inferSelect;

export async function processTodoExportJob(
  database: DbLike,
  input: { limit?: number; providerGroupId?: string | null } = {},
  options: {
    enabled?: boolean;
    webhookUrl?: string | null;
    timeoutSeconds?: number;
    httpClient?: HttpClient;
  } = {},
): Promise<number> {
  const settings = getSettings();
  const enabled = options.enabled ?? settings.TODO_EXPORT_ENABLED;
  if (!enabled) {
    logger.info("todo_export_disabled");
    return 0;
  }

  const webhookUrl = options.webhookUrl ?? settings.TODO_EXPORT_WEBHOOK_URL ?? null;
  if (!webhookUrl) {
    logger.warn("todo_export_webhook_missing");
    return 0;
  }

  const limit = Math.max(1, Math.min(input.limit ?? 100, 1000));
  const filters = [
    isNull(todos.exportedAt),
    inArray(todos.status, ["open", "in_progress"]),
    ...(input.providerGroupId ? [eq(todos.providerGroupId, input.providerGroupId)] : []),
  ];
  const rows = await database
    .select()
    .from(todos)
    .where(and(...filters))
    .orderBy(asc(todos.createdAt))
    .limit(limit);

  let exportedCount = 0;
  for (const todo of rows) {
    await database
      .update(todos)
      .set({
        exportAttemptCount: todo.exportAttemptCount + 1,
        updatedAt: new Date(),
      })
      .where(eq(todos.id, todo.id));

    try {
      const externalRef = await postTodo(webhookUrl, todo, {
        timeoutSeconds: options.timeoutSeconds ?? settings.TODO_EXPORT_TIMEOUT_SECONDS,
        httpClient: options.httpClient ?? fetch,
      });
      await database
        .update(todos)
        .set({
          exportedAt: new Date(),
          externalRef,
          lastExportError: null,
          updatedAt: new Date(),
        })
        .where(eq(todos.id, todo.id));
      exportedCount += 1;
    } catch (error) {
      const message = error instanceof Error ? error.message : String(error);
      await database
        .update(todos)
        .set({
          lastExportError: message.slice(0, 1000),
          updatedAt: new Date(),
        })
        .where(eq(todos.id, todo.id));
      logger.warn("todo_export_failed", {
        todo_id: todo.id,
        provider_group_id: todo.providerGroupId,
        error: message,
      });
    }
  }

  return exportedCount;
}

async function postTodo(
  webhookUrl: string,
  todo: TodoRow,
  options: { timeoutSeconds: number; httpClient: HttpClient },
): Promise<string | null> {
  const response = await options.httpClient(webhookUrl, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({
      id: todo.id,
      provider_group_id: todo.providerGroupId,
      message_id: todo.messageId,
      agent_run_id: todo.agentRunId,
      title: todo.title,
      description: todo.description,
      status: todo.status,
      priority: todo.priority,
      due_at: todo.dueAt?.toISOString() ?? null,
      created_at: todo.createdAt.toISOString(),
      updated_at: todo.updatedAt.toISOString(),
    }),
    signal: AbortSignal.timeout(options.timeoutSeconds * 1000),
  });

  if (!response.ok) {
    throw new Error(`todo_export_http_${response.status}: ${(await response.text()).slice(0, 300)}`);
  }

  const text = await response.text();
  if (!text.trim()) return null;

  const payload = JSON.parse(text) as unknown;
  if (payload && typeof payload === "object" && !Array.isArray(payload)) {
    const record = payload as Record<string, unknown>;
    const externalRef = record.external_ref ?? record.id;
    if (typeof externalRef === "string" && externalRef) {
      return externalRef;
    }
  }
  return null;
}
